### ISSUE TITLE
Add automated testing to verify correctness of lint fixes

### ISSUE THREAD

**Issue by barrywhart** [sqlfluff/sqlfluff#813]
As the number of SQLFluff rules with automated fixes increases, it becomes more and more important that these fixes **never, ever break or change the behavior** of existing code.

Here's a possible approach. Each rule has a `.yml` file with test cases, including the expected fix.
* For each test with a `fix_string`, introspect the "before" query to deduce a SQL schema that this query can actually be _executed_ against. The techniques for doing so will somewhat resemble those in #380, which determines which sources are CTEs and which are actual tables. Some type inference will be necessary for some queries, e.g. look at the operations or functions used on each column to determine possible types.
* Use a property-based testing tool like [Hypothesis](https://hypothesis.readthedocs.io/en/latest/) to generate test data for the tables. Note that this tool can generate complex test cases and narrow down issues to a "minimum" test case that causes an issue.
* Run both the "before" and "after" queries against the table and verify the results are identical.

Notes:
* I think something like this is vital before we can move SQLFluff from "Alpha" or "Beta" to "Production - Stable".
* The above approach involves some potentially challenging tool development. An alternative would be to manually develop test cases for each rule, but I think that will prove pretty onerous and not as effective.

**Comment by alanmcruickshank** [comment]
I agree we need something like this, but I'd assumed we would do something a little different. What do you think of this approach?

1. Pre `fix` we generate a parse tree of the existing query.
2. Post `fix` we generate a parse tree of the new query.
3. These trees are then compared in a "fuzzy" way to make sure they're "equivalent". Settling on these equivalence rules would probably be the tricky part, but could be defined for each `Segment` individually if required. On a broad scale I assume equivalence would mean:
    - Ignore any whitespace or comments
    - Ignore any aliasing
    - Everything else should be the same, but most of the time, the order is ignored.

If the equivalence fails, then the fix is not applied. Thoughts?

**Comment by barrywhart** [comment]
Both approaches are valid. I think they'll have differing strengths. Comparing parse trees will be great for checking rules that mainly reformat code.

I have a thought that certain fixes may be subject to causing subtle differences in behavior. For example, there was recently a rule added which replaces unnecessary CASE statements with simple Boolean expressions. That's essentially a rewrite of the code. At one point during the review cycle, I noticed a bug where it would apply the fix when it shouldn't, and that's what inspired this issue.

TL;DR: I see value in both, and I have a hard time deciding the priority of either. 😅
